[tp] improve SequenceParallel and its documentation

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131346

SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

cc @XilunWu @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wanchaol committed:
[tp] improve SequenceParallel and its documentation

SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

[ghstack-poisoned]

wanchaol committed:
Update on "[tp] improve SequenceParallel and its documentation"


SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -287,7 +287,10 @@ class SequenceParallel(ParallelStyle):
 
-    Both the input and output of the ``nn.Module`` will be sharded on the sequence dimension.
+    If the input passed in to this ``nn.Module`` is a :class:`torch.Tensor`, it assumes the input already sharded on the
+    sequence dimension and convert the input to a :class:`DTensor` sharded on the sequence dimension. If input passed in
```
nit: 
```suggestion
    sequence dimension and converts the input to a :class:`DTensor` sharded on the sequence dimension. If input passed in
```

awgu replied:
```suggestion
    sequence dimension and convert the input to a :class:`DTensor` sharded on the sequence dimension. If the input passed in
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -287,7 +287,10 @@ class SequenceParallel(ParallelStyle):
     `Reducing Activation Recomputation in Large Transformer Models <https://arxiv.org/abs/2205.05198>`__
 
-    Both the input and output of the ``nn.Module`` will be sharded on the sequence dimension.
+    If the input passed in to this ``nn.Module`` is a :class:`torch.Tensor`, it assumes the input already sharded on the
```
nit:
```suggestion
    If the input passed in to this ``nn.Module`` is a :class:`torch.Tensor`, it assumes that the input is already sharded on the
```

awgu commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -287,7 +287,10 @@ class SequenceParallel(ParallelStyle):
-    Both the input and output of the ``nn.Module`` will be sharded on the sequence dimension.
+    If the input passed in to this ``nn.Module`` is a :class:`torch.Tensor`, it assumes the input already sharded on the
+    sequence dimension and convert the input to a :class:`DTensor` sharded on the sequence dimension. If input passed in
+    to this ``nn.Module`` is already a :class:`DTensor` but not sharded on the sequence dimension, it would redistribute
```
```suggestion
    to this ``nn.Module`` is already a :class:`DTensor` but is not sharded on the sequence dimension, it would redistribute
```

wanchaol committed:
Update on "[tp] improve SequenceParallel and its documentation"


SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu reviewed (approved):
just some grammatical nits

wanchaol committed:
Update on "[tp] improve SequenceParallel and its documentation"


SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol committed:
Update on "[tp] improve SequenceParallel and its documentation"


SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol committed:
Update on "[tp] improve SequenceParallel and its documentation"


SequenceParallel style assumes the input torch.Tensor ALREADY sharded on
the sequence dimension if not passing in DTensor. Since it causes some
user confusion on the documentation, this PR:

1. for the case where input passed in is already a DTensor, we check the
   input placements and redistribute if it's not sharded on the sequence
dimension
2. update the doc to make it more explicit about the case when user
   passed in a torch.Tensor and DTensor

This would fix https://github.com/pytorch/pytorch/issues/129355

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -f "queued rocm jobs"
